### PR TITLE
add upstream host and cluster in StreamInfo interface for golang extension

### PR DESCRIPTION
**Pull request by StarryVae**
Commit Message: add upstream host and cluster in StreamInfo interface for golang extension
Additional Description: N/A
Risk Level: low
Testing: integration test
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by StarryVae** [comment]
@doujiang24 PTAL, thanks.

File: contrib/golang/filters/http/source/go/pkg/api/filter.go [hunk 1]
```diff
@@ -73,6 +73,10 @@ type StreamInfo interface {
 	DownstreamLocalAddress() string
 	// DownstreamRemoteAddress return the downstream remote address.
 	DownstreamRemoteAddress() string
+	// UpstreamHost return the upstream host address.
+	UpstreamHost() (string, bool)
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/api/filter.go:77]: `UpstreamHostAddress` could be better?

File: contrib/golang/filters/http/source/go/pkg/api/filter.go [hunk 2]
```diff
@@ -73,6 +73,10 @@ type StreamInfo interface {
...
+	// UpstreamHost return the upstream host cluster.
+	UpstreamCluster() (string, bool)
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/go/pkg/api/filter.go:79]: `UpstreamClusterName` could be better?

**Review by doujiang24 — COMMENTED** [review]
clear names could be better. Otherwise, LGTM.

File: contrib/golang/filters/http/source/go/pkg/api/filter.go (hunk 1 above)
Comment (StarryVae) [reply contrib/golang/filters/http/source/go/pkg/api/filter.go]: done, thanks.

File: contrib/golang/filters/http/source/go/pkg/api/filter.go (hunk 2 above)
Comment (StarryVae) [reply contrib/golang/filters/http/source/go/pkg/api/filter.go]: done, thanks.

**Comment by StarryVae** [comment]
And, i have a question, i try to run go example locally, i build the example first using:
`go build -o simple.so -buildmode=c-shared .`
then i run contrib envoy compiled locally with `simple.so`, then
`curl 127.0.0.1:10000/localreply_by_config`
finally, it will call `f.callbacks.SendLocalReply`, but, envoy crushs in `f.callbacks.SendLocalReply`.

i tried to figure out, but i have no idea why it crush 🤔

**Review by doujiang24 — DISMISSED** [review]

**Comment by doujiang24** [comment]
> And, i have a question, i try to run go example locally, i build the example first using: `go build -o simple.so -buildmode=c-shared .` then i run contrib envoy compiled locally with `simple.so`, then `curl 127.0.0.1:10000/localreply_by_config` finally, it will call `f.callbacks.SendLocalReply`, but, envoy crushs in `f.callbacks.SendLocalReply`.
> 
> i tried to figure out, but i have no idea why it crush 🤔 cc @doujiang24

@StarryVae Did you missing the `GODEBUG=cgocheck=0`?
see: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/golang_filter

**Comment by doujiang24** [comment]
/assign @wbpcode 
Thanks.

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 3]
```diff
@@ -910,6 +910,20 @@ CAPIStatus Filter::getStringValue(int id, GoString* value_str) {
...
+  case EnvoyValue::UpstreamHostAddress:
+    if (state.streamInfo().upstreamInfo() && state.streamInfo().upstreamInfo()->upstreamHost()) {
+      req_->strValue = state.streamInfo().upstreamInfo()->upstreamHost()->address()->asString();
+    } else {
+      return CAPIStatus::CAPIValueNotFound;
+    }
+    break;
+  case EnvoyValue::UpstreamClusterName:
+    if (state.streamInfo().upstreamInfo() && state.streamInfo().upstreamInfo()->upstreamHost()) {
+      req_->strValue = state.streamInfo().upstreamInfo()->upstreamHost()->cluster().name();
```
Comment (wbpcode) [inline contrib/golang/filters/http/source/golang_filter.cc:922]: Try to get cluster info from stream info directly. It's possible that cluster is selected but no upstream host.

Comment (StarryVae) [reply contrib/golang/filters/http/source/golang_filter.cc]: done, thanks.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by StarryVae** [comment]
> > And, i have a question, i try to run go example locally, i build the example first using: `go build -o simple.so -buildmode=c-shared .` then i run contrib envoy compiled locally with `simple.so`, then `curl 127.0.0.1:10000/localreply_by_config` finally, it will call `f.callbacks.SendLocalReply`, but, envoy crushs in `f.callbacks.SendLocalReply`.
> > i tried to figure out, but i have no idea why it crush 🤔 cc @doujiang24
> 
> @StarryVae Did you missing the `GODEBUG=cgocheck=0`? see: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/golang_filter
> 
> if it's not the `cgocheck` issue, please create an issue with the backtrace error log, thanks.

@doujiang24 yeah, it is the `cgocheck` issue, thanks.

**Pull request merged**